File: moji projekti/Web SCreper/web_scraper.js
const { chromium } = require('playwright');
const fs = require('fs');

class RealEstateScraper {
    constructor() {
        this.resultsBelgrade = [];
        this.resultsSerbia = [];
        this.outputFileBelgrade = '/home/dju/moji projekti/Web SCreper/rezultati_beograd.txt';
        this.outputFileSerbia = '/home/dju/moji projekti/Web SCreper/rezultati_srbija.txt';
    }

    saveResultsBelgrade() {
        let content = `REZULTATI ZA BEOGRAD I OKOLINU\n`;
        content += `Datum: ${new Date().toLocaleString()}\n`;
        content += `Kriterijumi: Kuće i seoska domaćinstva, 15.000 - 20.000 EUR, Beograd i okolina\n`;
        content += `${'='.repeat(80)}\n\n`;

        if (this.resultsBelgrade.length === 0) {
            content += 'Nema pronađenih oglasa u zadatom kriterijumu.\n\n';
            content += 'Direktni linkovi za pretragu KUĆA:\n';
            content += '• https://www.nekretnine.rs/stambeni-objekti/kuce/beograd/cena/15000_20000/\n';
            content += '• https://www.olx.rs/d/nedviznosti/prodazha/kukhi/beograd/\n';
            content += '• https://www.kupujemprodajem.com/nekretnine/kuce/beograd/\n\n';
            content += 'Direktni linkovi za pretragu SEOSKIH DOMAĆINSTAVA:\n';
            content += '• https://www.nekretnine.rs/stambeni-objekti/seoska-domacinstva/beograd/cena/15000_20000/\n';
            content += '• https://www.olx.rs/d/nedviznosti/prodazha/seosko-domacinstvo/beograd/\n';
            content += '• https://www.kupujemprodajem.com/nekretnine/seoska-domacinstva/beograd/\n';
        } else {
            content += `UKUPNO PRONAĐENO: ${this.resultsBelgrade.length} oglasa\n\n`;
            this.resultsBelgrade.forEach((ad, index) => {
                content += `${index + 1}. ${ad.title}\n`;
                content += `   Cena: ${ad.price}\n`;
                content += `   Lokacija: ${ad.location}\n`;
                content += `   Tip: ${ad.type || 'N/A'}\n`;
                content += `   Link: ${ad.link}\n`;
                content += `   Izvor: ${ad.source}\n`;
                content += `${'-'.repeat(80)}\n\n`;
            });
        }

        content += `\n${'='.repeat(80)}\n`;
        content += 'NAPOMENA: Ovo su rezultati automatske pretrage.\n';
        content += 'Preporučuje se ručna provera svakog oglasa.\n';
        content += `${'='.repeat(80)}\n`;

        fs.writeFileSync(this.outputFileBelgrade, content, 'utf8');
        console.log(`\n✓ Rezultati za Beograd sačuvani u: ${this.outputFileBelgrade}`);
    }

    saveResultsSerbia() {
        let content = `REZULTATI ZA ČITAVU SRBIJU\n`;
        content += `Datum: ${new Date().toLocaleString()}\n`;
        content += `Kriterijumi: Kuće i seoska domaćinstva, 15.000 - 20.000 EUR, Srbija\n`;
        content += `${'='.repeat(80)}\n\n`;

        if (this.resultsSerbia.length === 0) {
            content += 'Nema pronađenih oglasa u zadatom kriterijumu.\n\n';
            content += 'Direktni linkovi za pretragu KUĆA:\n';
            content += '• https://www.nekretnine.rs/stambeni-objekti/kuce/srbija/cena/15000_20000/\n';
            content += '• https://www.olx.rs/d/nedviznosti/prodazha/kukhi/\n';
            content += '• https://www.kupujemprodajem.com/nekretnine/kuce/\n\n';
            content += 'Direktni linkovi za pretragu SEOSKIH DOMAĆINSTAVA:\n';
            content += '• https://www.nekretnine.rs/stambeni-objekti/seoska-domacinstva/srbija/cena/15000_20000/\n';
            content += '• https://www.olx.rs/d/nedviznosti/prodazha/seosko-domacinstvo/\n';
            content += '• https://www.kupujemprodajem.com/nekretnine/seoska-domacinstva/\n';
        } else {
            content += `UKUPNO PRONAĐENO: ${this.resultsSerbia.length} oglasa\n\n`;
            this.resultsSerbia.forEach((ad, index) => {
                content += `${index + 1}. ${ad.title}\n`;
                content += `   Cena: ${ad.price}\n`;
                content += `   Lokacija: ${ad.location}\n`;
                content += `   Tip: ${ad.type || 'N/A'}\n`;
                content += `   Link: ${ad.link}\n`;
                content += `   Izvor: ${ad.source}\n`;
                content += `${'-'.repeat(80)}\n\n`;
            });
        }

        content += `\n${'='.repeat(80)}\n`;
        content += 'NAPOMENA: Ovo su rezultati automatske pretrage.\n';
        content += 'Preporučuje se ručna provera svakog oglasa.\n';
        content += `${'='.repeat(80)}\n`;

        fs.writeFileSync(this.outputFileSerbia, content, 'utf8');
        console.log(`\n✓ Rezultati za Srbiju sačuvani u: ${this.outputFileSerbia}`);
    }

    extractPrice(priceText) {
        if (!priceText) return null;
        const match = priceText.match(/(\d[\d.,\s]*)/);
        if (match) {
            const priceNum = parseInt(match[1].replace(/[^\d]/g, ''));
            if (priceNum >= 15000 && priceNum <= 20000) {
                return priceNum;
            }
        }
        return null;
    }

    async extractAdsFromPage(page, selector, source) {
        return await page.$$eval(selector, (items, src) => {
            return items.map(item => {
                const titleEl = item.querySelector('h2, h3, h6, .offer-title, .property-title, .adTitle, [data-testid="ad-title"], .title, a');
                const priceEl = item.querySelector('.offer-price, .property-price, .adPrice, [data-testid="ad-price"], .price');
                const locationEl = item.querySelector('.offer-location, .property-location, .adLocation, [data-testid="location-date"], [data-testid="ad-location"], .location');
                const linkEl = item.querySelector('a');

                return {
                    title: titleEl ? titleEl.innerText.trim() : 'Nema naslova',
                    price: priceEl ? priceEl.innerText.trim() : 'N/A',
                    location: locationEl ? locationEl.innerText.trim() : 'N/A',
                    link: linkEl ? linkEl.href : 'N/A',
                    source: src
                };
            });
        }, source);
    }

    // === BEOGRAD PRETRAGE ===

    async searchNekretnineRsBelgrade(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem nekretnine.rs - KUĆE...');
        try {
            await page.goto('https://www.nekretnine.rs/stambeni-objekti/kuce/beograd/cena/15000_20000/', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.offer-item, .property-item, [data-testid="listing-item"]', 'nekretnine.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchNekretnineRsBelgradeDomacinstva(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem nekretnine.rs - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.nekretnine.rs/stambeni-objekti/seoska-domacinstva/beograd/cena/15000_20000/', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.offer-item, .property-item, [data-testid="listing-item"]', 'nekretnine.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchOlxRsBelgrade(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem OLX.rs - KUĆE...');
        try {
            await page.goto('https://www.olx.rs/d/nedviznosti/prodazha/kukhi/beograd/', {
                waitUntil: 'domcontentloaded',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '[data-cy="l-card"], .css-qfzx1y, [data-testid="listing-ad"]', 'olx.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchOlxRsBelgradeDomacinstva(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem OLX.rs - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.olx.rs/d/nedviznosti/prodazha/seosko-domacinstvo/beograd/', {
                waitUntil: 'domcontentloaded',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '[data-cy="l-card"], .css-qfzx1y, [data-testid="listing-ad"]', 'olx.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchKupujemProdajemBelgrade(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem KupujemProdajem.com - KUĆE...');
        try {
            await page.goto('https://www.kupujemprodajem.com/nekretnine/kuce/beograd/cena/15000/20000', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.adItem, [data-testid="ad-list-item"], .AdItem', 'kupujemprodajem.com');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchKupujemProdajemBelgradeDomacinstva(page) {
        console.log('\n📍 [BEOGRAD] Pretražujem KupujemProdajem.com - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.kupujemprodajem.com/nekretnine/seoska-domacinstva/beograd/cena/15000/20000', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.adItem, [data-testid="ad-list-item"], .AdItem', 'kupujemprodajem.com');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsBelgrade.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    // === SRBIJA PRETRAGE ===

    async searchNekretnineRsSerbia(page) {
        console.log('\n📍 [SRBIJA] Pretražujem nekretnine.rs - KUĆE...');
        try {
            await page.goto('https://www.nekretnine.rs/stambeni-objekti/kuce/srbija/cena/15000_20000/', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.offer-item, .property-item, [data-testid="listing-item"]', 'nekretnine.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchNekretnineRsSerbiaDomacinstva(page) {
        console.log('\n📍 [SRBIJA] Pretražujem nekretnine.rs - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.nekretnine.rs/stambeni-objekti/seoska-domacinstva/srbija/cena/15000_20000/', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.offer-item, .property-item, [data-testid="listing-item"]', 'nekretnine.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchOlxRsSerbia(page) {
        console.log('\n📍 [SRBIJA] Pretražujem OLX.rs - KUĆE...');
        try {
            await page.goto('https://www.olx.rs/d/nedviznosti/prodazha/kukhi/', {
                waitUntil: 'domcontentloaded',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '[data-cy="l-card"], .css-qfzx1y, [data-testid="listing-ad"]', 'olx.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchOlxRsSerbiaDomacinstva(page) {
        console.log('\n📍 [SRBIJA] Pretražujem OLX.rs - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.olx.rs/d/nedviznosti/prodazha/seosko-domacinstvo/', {
                waitUntil: 'domcontentloaded',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '[data-cy="l-card"], .css-qfzx1y, [data-testid="listing-ad"]', 'olx.rs');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchKupujemProdajemSerbia(page) {
        console.log('\n📍 [SRBIJA] Pretražujem KupujemProdajem.com - KUĆE...');
        try {
            await page.goto('https://www.kupujemprodajem.com/nekretnine/kuce/cena/15000/20000', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.adItem, [data-testid="ad-list-item"], .AdItem', 'kupujemprodajem.com');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Kuća' });
                    console.log(`  ✓ Pronađena kuća: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async searchKupujemProdajemSerbiaDomacinstva(page) {
        console.log('\n📍 [SRBIJA] Pretražujem KupujemProdajem.com - SEOSKA DOMAĆINSTVA...');
        try {
            await page.goto('https://www.kupujemprodajem.com/nekretnine/seoska-domacinstva/cena/15000/20000', {
                waitUntil: 'networkidle',
                timeout: 30000
            });
            await page.waitForTimeout(3000);

            const ads = await this.extractAdsFromPage(page, '.adItem, [data-testid="ad-list-item"], .AdItem', 'kupujemprodajem.com');

            ads.forEach(ad => {
                const priceNum = this.extractPrice(ad.price);
                if (priceNum) {
                    this.resultsSerbia.push({ ...ad, type: 'Seosko domaćinstvo' });
                    console.log(`  ✓ Pronađeno domaćinstvo: ${ad.title.substring(0, 50)}... (${ad.price})`);
                }
            });
        } catch (error) {
            console.log(`  ⚠ Greška: ${error.message}`);
        }
    }

    async run() {
        console.log('\n' + '='.repeat(80));
        console.log('WEB SCRAPER - Kuće i seoska domaćinstva');
        console.log('Cilj: 15.000 - 20.000 EUR');
        console.log('='.repeat(80));

        const browser = await chromium.launch({ headless: true });
        const context = await browser.newContext({
            userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
        });
        const page = await context.newPage();

        // === BEOGRAD PRETRAGA ===
        console.log('\n\n' + '🏠'.repeat(20));
        console.log('PRETRAGA ZA BEOGRAD I OKOLINU');
        console.log('🏠'.repeat(20) + '\n');
        
        await this.searchNekretnineRsBelgrade(page);
        await this.searchNekretnineRsBelgradeDomacinstva(page);
        await this.searchOlxRsBelgrade(page);
        await this.searchOlxRsBelgradeDomacinstva(page);
        await this.searchKupujemProdajemBelgrade(page);
        await this.searchKupujemProdajemBelgradeDomacinstva(page);

        // === SRBIJA PRETRAGA ===
        console.log('\n\n' + '🇷🇸'.repeat(20));
        console.log('PRETRAGA ZA ČITAVU SRBIJU');
        console.log('🇷🇸'.repeat(20) + '\n');
        
        await this.searchNekretnineRsSerbia(page);
        await this.searchNekretnineRsSerbiaDomacinstva(page);
        await this.searchOlxRsSerbia(page);
        await this.searchOlxRsSerbiaDomacinstva(page);
        await this.searchKupujemProdajemSerbia(page);
        await this.searchKupujemProdajemSerbiaDomacinstva(page);

        await browser.close();

        // Čuva rezultate
        this.saveResultsBelgrade();
        this.saveResultsSerbia();

        // Finalni rezime
        console.log('\n\n' + '='.repeat(80));
        console.log('PRETRAGA ZAVRŠENA!');
        console.log('='.repeat(80));
        console.log(`\n📊 REZULTATI:`);
        console.log(`   • Beograd i okolina: ${this.resultsBelgrade.length} oglasa`);
        console.log(`     → Fajl: ${this.outputFileBelgrade}`);
        console.log(`   • Srbija (cela): ${this.resultsSerbia.length} oglasa`);
        console.log(`     → Fajl: ${this.outputFileSerbia}`);
        console.log(`\n📝 UKUPNO: ${this.resultsBelgrade.length + this.resultsSerbia.length} oglasa`);
        console.log(`\n🔍 Pretražene kategorije:`);
        console.log(`   • Kuće`);
        console.log(`   • Seoska domaćinstva`);
        console.log('='.repeat(80) + '\n');
    }
}

// Pokreće scraper
const scraper = new RealEstateScraper();
scraper.run().catch(error => {
    console.error('Greška:', error);
    process.exit(1);
});

File: moji projekti/Web SCreper/web_scraper.py
#!/usr/bin/env python3
"""
Web Scraper za seoska imanja u okolini Beograda
Cilj: Pronalazak oglasa od 15.000 - 20.000 eura
"""

import asyncio
import json
import re
from datetime import datetime
from playwright.async_api import async_playwright
from urllib.parse import urljoin, quote

class RealEstateScraper:
    def __init__(self):
        self.results = []
        self.output_file = "/home/dju/moji projekti/Web SCreper/rezultati_oglasi.txt"
        
    def save_results(self):
        """Čuva rezultate u tekstualni fajl"""
        with open(self.output_file, 'w', encoding='utf-8') as f:
            f.write(f"Rezultati pretrage seoskih imanja\n")
            f.write(f"Datum: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")
            f.write(f"Kriterijumi: 15.000 - 20.000 EUR, okolina Beograda\n")
            f.write("=" * 80 + "\n\n")
            
            if not self.results:
                f.write("Nema pronađenih oglasa.\n")
            else:
                for idx, ad in enumerate(self.results, 1):
                    f.write(f"{idx}. {ad.get('title', 'Nema naslova')}\n")
                    f.write(f"   Cena: {ad.get('price', 'N/A')}\n")
                    f.write(f"   Lokacija: {ad.get('location', 'N/A')}\n")
                    f.write(f"   Opis: {ad.get('description', 'N/A')}\n")
                    f.write(f"   Link: {ad.get('link', 'N/A')}\n")
                    f.write("-" * 80 + "\n\n")
                    
        print(f"Rezultati sačuvani u: {self.output_file}")
        
    def is_price_in_range(self, price_text):
        """Proverava da li je cena u rasponu 15.000 - 20.000 EUR"""
        if not price_text:
            return False
            
        # Uklanja razmake i ekstrahuje brojeve
        numbers = re.findall(r'\d+', price_text.replace('.', '').replace(',', ''))
        if numbers:
            try:
                price = int(numbers[0])
                # Proverava da li je cena između 15.000 i 20.000
                return 15000 <= price <= 20000
            except ValueError:
                pass
        return False
        
    def is_near_belgrade(self, location):
        """Proverava da li je lokacija u okolini Beograda"""
        if not location:
            return False
            
        location_lower = location.lower()
        belgrade_keywords = [
            'beograd', 'bg', 'beo', 'grad', 'novi beograd', 'zemun', 'rakovica',
            'voždovac', 'voždovač', 'čukarica', 'čukarič', 'palilula', 'zvezdara',
            'grocka', 'barajevo', 'lazarevac', 'mladenovac', 'obrenovac', 'sopot',
            'surčin', 'surcin', 'zemun', 'zemunsk',
            'batajnica', 'borča', 'borca', 'kotež', 'kotez', 'umka', 'rušanj', 'rusanj',
            'ripanj', 'jajinci', 'mirijevo', 'kotež', 'leštane', 'lestane',
            'kaluđerica', 'kaluderica', 'vinča', 'vinca', 'padinska skela'
        ]
        
        return any(keyword in location_lower for keyword in belgrade_keywords)

    async def search_nekretnine_rs(self, page):
        """Pretražuje nekretnine.rs"""
        print("Pretražujem nekretnine.rs...")
        try:
            # URL za pretragu zemljišta kuća u Beogradu
            url = "https://www.nekretnine.rs/stambeni-objekti/kuce/beograd/cena/15000_20000/"
            await page.goto(url, wait_until='networkidle', timeout=30000)
            
            # Čeka da se oglasi učitaju
            await page.wait_for_selector('.offer-title, .property-title, h2', timeout=10000)
            
            # Prikupi oglase
            ads = await page.query_selector_all('.offer-item, .property-item, .oglas, article')
            
            for ad in ads:
                try:
                    title_elem = await ad.query_selector('h2, .offer-title, .property-title, a')
                    price_elem = await ad.query_selector('.offer-price, .property-price, .price')
                    location_elem = await ad.query_selector('.offer-location, .property-location, .location')
                    desc_elem = await ad.query_selector('.offer-description, .property-description, .description')
                    link_elem = await ad.query_selector('a')
                    
                    title = await title_elem.inner_text() if title_elem else "Nema naslova"
                    price = await price_elem.inner_text() if price_elem else "N/A"
                    location = await location_elem.inner_text() if location_elem else "N/A"
                    description = await desc_elem.inner_text() if desc_elem else "N/A"
                    link = await link_elem.get_attribute('href') if link_elem else "N/A"
                    
                    if link and link.startswith('/'):
                        link = urljoin("https://www.nekretnine.rs", link)
                    
                    # Proverava kriterijume
                    if self.is_price_in_range(price) and self.is_near_belgrade(location):
                        self.results.append({
                            'title': title.strip(),
                            'price': price.strip(),
                            'location': location.strip(),
                            'description': description.strip()[:200] + "..." if len(description) > 200 else description.strip(),
                            'link': link,
                            'source': 'nekretnine.rs'
                        })
                        print(f"✓ Pronađen oglas: {title[:50]}...")
                        
                except Exception as e:
                    continue
                    
        except Exception as e:
            print(f"Greška pri pretrazi nekretnine.rs: {e}")

    async def search_olx_rs(self, page):
        """Pretražuje OLX.rs"""
        print("Pretražujem OLX.rs...")
        try:
            # URL za pretragu zemljišta kuća u Beogradu
            url = "https://www.olx.rs/d/nedviznosti/prodazha/kukhi/beograd/q-ku%C4%87a/"
            await page.goto(url, wait_until='networkidle', timeout=30000)
            
            # Čeka da se oglasi učitaju
            await page.wait_for_selector('.css-qfzx1y, [data-testid="listing-ad"]', timeout=10000)
            
            # Prikupi oglase
            ads = await page.query_selector_all('.css-qfzx1y, [data-testid="listing-ad"]')
            
            for ad in ads:
                try:
                    title_elem = await ad.query_selector('h6, .css-16v5mdi, [data-testid="ad-title"]')
                    price_elem = await ad.query_selector('.css-10b0gli, [data-testid="ad-price"]')
                    location_elem = await ad.query_selector('.css-ve6ph5, [data-testid="location-date"]')
                    link_elem = await ad.query_selector('a')
                    
                    title = await title_elem.inner_text() if title_elem else "Nema naslova"
                    price = await price_elem.inner_text() if price_elem else "N/A"
                    location = await location_elem.inner_text() if location_elem else "N/A"
                    link = await link_elem.get_attribute('href') if link_elem else "N/A"
                    
                    if link and link.startswith('/'):
                        link = urljoin("https://www.olx.rs", link)
                    
                    # Proverava kriterijume
                    if self.is_price_in_range(price) and self.is_near_belgrade(location):
                        self.results.append({
                            'title': title.strip(),
                            'price': price.strip(),
                            'location': location.strip(),
                            'description': "Pogledajte detalje na linku",
                            'link': link,
                            'source': 'olx.rs'
                        })
                        print(f"✓ Pronađen oglas: {title[:50]}...")
                        
                except Exception as e:
                    continue
                    
        except Exception as e:
            print(f"Greška pri pretrazi OLX.rs: {e}")

    async def search_halooglasi(self, page):
        """Pretražuje Halooglasi"""
        print("Pretražujem Halooglasi...")
        try:
            # URL za pretragu kuća u Beogradu
            url = "https://www.halooglasi.com/nedviznosti/prodaja-kukha/beograd"
            await page.goto(url, wait_until='networkidle', timeout=30000)
            
            # Čeka da se oglasi učitaju
            await page.wait_for_selector('.product-item, .listing-item', timeout=10000)
            
            # Prikupi oglase
            ads = await page.query_selector_all('.product-item, .listing-item')
            
            for ad in ads:
                try:
                    title_elem = await ad.query_selector('.product-title, h3, a')
                    price_elem = await ad.query_selector('.product-price, .price')
                    location_elem = await ad.query_selector('.product-location, .location')
                    link_elem = await ad.query_selector('a')
                    
                    title = await title_elem.inner_text() if title_elem else "Nema naslova"
                    price = await price_elem.inner_text() if price_elem else "N/A"
                    location = await location_elem.inner_text() if location_elem else "N/A"
                    link = await link_elem.get_attribute('href') if link_elem else "N/A"
                    
                    if link and link.startswith('/'):
                        link = urljoin("https://www.halooglasi.com", link)
                    
                    # Proverava kriterijume
                    if self.is_price_in_range(price) and self.is_near_belgrade(location):
                        self.results.append({
                            'title': title.strip(),
                            'price': price.strip(),
                            'location': location.strip(),
                            'description': "Pogledajte detalje na linku",
                            'link': link,
                            'source': 'halooglasi'
                        })
                        print(f"✓ Pronađen oglas: {title[:50]}...")
                        
                except Exception as e:
                    continue
                    
        except Exception as e:
            print(f"Greška pri pretrazi Halooglasi: {e}")

    async def run(self):
        """Pokreće scraper na svim sajtovima"""
        print("\n" + "=" * 80)
        print("POKRETANJE WEB SCRAPER-A ZA SEOSKA IMANJA")
        print("=" * 80 + "\n")
        
        async with async_playwright() as p:
            # Pokreće browser
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context(
                user_agent='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
            )
            page = await context.new_page()
            
            # Pretražuje sve sajtove
            await self.search_nekretnine_rs(page)
            await self.search_olx_rs(page)
            await self.search_halooglasi(page)
            
            # Zatvara browser
            await browser.close()
        
        # Čuva rezultate
        self.save_results()
        
        # Prikazuje rezime
        print("\n" + "=" * 80)
        print(f"PRETRAGA ZAVRŠENA!")
        print(f"Ukupno pronađenih oglasa: {len(self.results)}")
        print(f"Rezultati sačuvani u: {self.output_file}")
        print("=" * 80 + "\n")

if __name__ == "__main__":
    scraper = RealEstateScraper()
    asyncio.run(scraper.run())
